### PR TITLE
persist: implement compaction v1

### PR DESCRIPTION
**Pull request by ruchirK**
~This pr is not ready for merging and only exists to talk about the approach!~

~This commit introduces a prototype compaction function which attempts to compact a
trace up to a `as_of` time, but instead of compacting all updates <= as_of, it merely
finds the subrange of batches containing times < as_of and compacts them into a single
batch and forwards their updates to `max(batch.upper()) - 1`.~

Future work:

* ~Actually compact up to as_of (this requires potentially splitting a batch in two)~ decided against doing this as differential does not do this.
* Push the computationally intensive work of reading batches and consolidating them onto
  a separate thread -- will do this in a separate commit
* ~Figure out a better interface between the Indexed, Trace, and Cache layers (right now we
  rediscover the Id for the Trace by looking inside a previously persisted batch)~ addressed in a separate commit

### CODE REVIEW

**Comment by ruchirK** [comment]
First three commits are #7446

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [inline src/persist/src/indexed/mod.rs:362]: move this check to the function on trace

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [inline src/persist/src/indexed/mod.rs:354]: change the argument to since

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [inline src/persist/src/indexed/mod.rs:353]: reword comment

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [inline src/persist/src/indexed/trace.rs:52]: replace todo with a note saying we can split up the physical and logical compaction parts at a later time

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [inline src/persist/src/indexed/trace.rs:185]: since frontier is closed

**Comment by ruchirK** [comment]
RFAL! The first commit does some preliminary wiring up and sets up `BlobTrace` to store a since frontier / save it to meta and wires up `allow_compaction` to advance that since frontier, and then the second commit actually does the work of defining a `merge` that can physically and logically compact two batches, and a `step` function that determines "whats the next pair of batches to merge, if any". The whole thing is set up to maintain a logarithmic number of batches / perform a logarithmic number of merges wrt the number of batches appended to the trace.

I'm going to go and add more test coverage to both commits now so pls hold off on reviewing the (lack of) test coverage.

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [inline src/persist/src/indexed/runtime.rs:240]: address this.

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [inline src/persist/src/indexed/trace.rs:93]: would love a second opinion on this one!

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [inline src/persist/src/indexed/trace.rs:221]: would love second opinions on the two todos in this function as well

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:223]: Can you put a big comment overview on how the compaction strategy works? Sorry if I missed this somewhere else

**Review by danhhz — COMMENTED** [review]
Didn't get a chance to look at this before I stopped for the day, I'll catch this tomorrow

File: src/persist/src/indexed/encoding.rs
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:358]: should this also be "at or in advance of"?

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:283]: Should this be an error or a no-op? I'm divided but a very slight lean toward no-op. What was your thinking here?

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:40]: nit: an append-only

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:42]: nit: defines a time

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:47]: "union of those intervals"?

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:76]: nit: "Specifically, compaction levels never increase." is a little confusing because you're talking about iterating from low to high indexes, but an individual batches compaction level does increase. I think we just need a few more words here to clarify

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:124]: Yes. Otherwise we get wildly different behavior after a restart no?

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:168]: Toss in a comment to the effect of "As mentioned in the big comment on BlobTrace, a batch's initial level is set to 1."

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:195]: Should this really be an error? On one hand, it's unclear what the output level would be, but on the other, this feels to me more like a debug assertion in the caller of merge or something. I could be convinced

Edit: Thought about this more as I finished the review and I think it's fine to leave as-is, but something about this feels like a wart to me

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:227]: nit: move the construction of this Description to be right below the `first.desc.upper() != second.desc.lower()` if block. the first/second lower/upper relationship is more clear there

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:223]: nit: remove this comment, it doesn't add anything

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:226]: nit: remove this comment, it doesn't add anything

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:240]: *eye twitches* :) (at the formatting, not the content)

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:262]: Seems like you want a linked list? I've never used it but there seems to be one in the stdlib

You could also do it in two steps where you remove a thing and replace a thing which at least is only one memcopy of everything in the vec

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:265]: somewhere around here is a pretty good place to debug_assert that validate is happy. speaking of which, we should add some validation stuff around the level invariants

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:254]: if we don't end up making this a linked list, you could use https://doc.rust-lang.org/std/slice/struct.Windows.html to do this iteration without doing index math

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:252]: this is just to avoid iterating over self.batches every time it's called? this one seems fine to leave as a TODO

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:393]: assert that the structure after you compact is what you expect

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:244]: nit: crate a `let merged_batch_level = first.level + 1` up by where you assert that the two input levels are the same and use that here. makes it more obvious where this is coming from

File: src/persist/src/indexed/encoding.rs
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:346]: yeah exactly -- the invariant is that since is either 0 or less than upper (seemed more compact than making it dependent on `if upper == 0`. can restructure the invariant if that helps

File: src/persist/src/indexed/encoding.rs
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:358]: `less_equal` bites again! I wrote this to say `if !(desc.since <= self.since)` which implies that `desc.since > self.since` (in one dimension). glad im not the only one getting tripped up by this :)

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:283]: My thought was that it would be nice if `allow_compaction`'s semantics matched `seal`'s semantics as much as possible (and perhaps that means we should go all the way and make this `less_equal` here rather than `less_than`)

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:195]: this could be a debug assert inside of merge?

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:127]: (this was a good nerd snipe) I spent a while thinking about this. I think you can kind of hand wave a proof that -- the more batches you have the more likely you are to have a sequence of 1 or two batches at each level

like if you've seen > 1 million batches go by its extremely unlikely that your list of levels will look like

`[20, 1, 1, 1, ,1 ... ]`

and its more likely that your list will look like

`[20, 19, 18, 16, 16, 15, 14, 13, .... 1, 1, 1, ...]`

and if that's true then I can vaguely hand-wave my way into saying "if you recast as 'everything is on the same level' then you wouldn't do more than 2x the optimal amount of work" because intuitively, you want to merge the last three elements when you have

`[20, 19, 19 ... ]`

which turns into 

`[20, 20]`...
 
 which turns into `[21, ... ]`
 
 but in some sense you haven't received enough batches yet to go down the tree to make another level 19, for which you need another level 18, for which you need two more level 17s, ... you need something approximately 2^18 more batches to "pay for" the final level compaction, but the suboptimality of this is (maybe probabilistically) bounded as a fraction over the optimal compaction strategy.
 
 Anyway, that was just a rant; I'll go ahead and serialize the whole thing.

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:257]: I think you still need the indices to know where to remove / insert from?

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs]: done and done

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs]: i placed debug asserts for "batches are the same level and contiguous"

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:236]: hmmm i think linked list would negatively affect the serialization to meta which is likely more perf sensitive than this step operation. I did the swap remove thing you asked so now the cost is one memcpy which seems like a fine place to stop in this pr and revisit if this ends up being a bottleneck

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:264]: I made allow_compaction's requirements as strict as seal's

**Comment by ruchirK** [comment]
Ok I've responded to all the comments - thanks for such a thorough review! This is ready for a second look :)

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:228]: .enumerate(), but yeah at that point it's less clear than doing it this way.

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:236]: Works for me!

File: src/persist/src/indexed/encoding.rs
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:129]: why don't the levels start counting from 0? if there's some reason for this that i'm not seeing, we should probably spell it out in a comment

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [inline src/persist/src/indexed/trace.rs:321]: Is this comment wrong?

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:359]: nit: empty line between paragraphs

File: src/persist/src/indexed/encoding.rs
Comment (danhhz) [reply src/persist/src/indexed/encoding.rs:364]: nope, this is fine. just checking my understanding

**Review by danhhz — APPROVED** [review]
Sorry it took so long to get back to this. It was hard to see the diff since the last version I reviewed and github keeps hiding my comments as outdated or resolved. I did manage to finally trick github into showing me the diff (https://github.com/MaterializeInc/materialize/compare/6f275d6..de02346), but I suspect this would all be much easier if non-trivial PRs had review comments addressed append-only until the LGTM. Let's try that going forward, please!

LGTM modulo the "first.desc.upper(), second.desc.lower()" comment

File: src/persist/src/indexed/encoding.rs
Comment (ruchirK) [reply src/persist/src/indexed/encoding.rs:129]: 🤦🏾 yeah ofc that makes more sense

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:321]: changed to be more clear

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:181]: makes sense and done

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:238]: changed to assign

**Comment by ruchirK** [comment]
TFTR! and will do on keeping review changes append only till the final approval -- sorry about that! merging on green

**Pull request merged**
